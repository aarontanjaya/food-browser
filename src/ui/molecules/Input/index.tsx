import { cn } from '@app/utils';
import React from 'react';

type InputProps = {
  'data-testid'?: string;
  LeftElement?: React.ReactNode;
  inputClassName?: string;
} & React.ComponentProps<'input'>;

export default function Input({ 'data-testid': testId, LeftElement, inputClassName, className, ...rest }: InputProps) {
  return (
    <div className={cn(`flex flex-col`, className)}>
      <div
        data-testid={`${testId}:container`}
        className={cn(
          'flex flex-1 flex-row items-center gap-x-3 rounded-md border border-neutral-200 bg-white p-2',
          inputClassName,
        )}
      >
        {LeftElement}
        <input
          data-testid={`${testId}:input`}
          className={'w-full bg-transparent placeholder-neutral-500 outline-0'}
          {...rest}
        />
      </div>
    </div>
  );
}
