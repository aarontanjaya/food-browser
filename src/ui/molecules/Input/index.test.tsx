import Input from '.';
import { fireEvent, render } from '@testing-library/react';
import { act } from 'react';

describe('Input', () => {
  it('should render correctly without error and have required components', () => {
    // #region ARRANGE
    const screen = render(<Input data-testid="input" />);
    // #endregion

    // #region ASSERT
    expect(screen).toMatchSnapshot();
    expect(() => screen.getByTestId('input:container')).not.toThrow();
    expect(() => screen.getByTestId('input:input')).not.toThrow();
    // #endregion
  });

  it('should render leftElement if provided', () => {
    // #region ARRANGE
    const screen = render(<Input LeftElement={<div data-testid="left-element" />} data-testid="input" />);
    // #endregion

    // #region ASSERT
    expect(screen).toMatchSnapshot();
    expect(() => screen.getByTestId('left-element')).not.toThrow();
    // #endregion
  });

  it('should call onChange when input value changes', () => {
    // #region ARRANGE
    const mockOnChange = jest.fn();
    const screen = render(<Input data-testid="input" onChange={mockOnChange} />);
    const input = screen.getByTestId('input:input');
    // #endregion

    // #region ACT
    act(() => {
      fireEvent.change(input, { target: { value: 'new value' } });
    });

    // #endregion

    // #region ASSERT
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(
      expect.objectContaining({ target: expect.objectContaining({ value: 'new value' }) }),
    );
    // #endregion
  });
});
