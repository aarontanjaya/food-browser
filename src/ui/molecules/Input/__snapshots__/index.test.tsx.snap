// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly without error and have required components 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col"
      >
        <div
          class="flex flex-1 flex-row items-center gap-x-3 rounded-xl border border-neutral-200 bg-white p-3 px-5"
          data-testid="input:container"
        >
          <input
            class="w-full bg-transparent placeholder-neutral-500 outline-0"
            data-testid="input:input"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col"
    >
      <div
        class="flex flex-1 flex-row items-center gap-x-3 rounded-xl border border-neutral-200 bg-white p-3 px-5"
        data-testid="input:container"
      >
        <input
          class="w-full bg-transparent placeholder-neutral-500 outline-0"
          data-testid="input:input"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input should render leftElement if provided 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col"
      >
        <div
          class="flex flex-1 flex-row items-center gap-x-3 rounded-xl border border-neutral-200 bg-white p-3 px-5"
          data-testid="input:container"
        >
          <div
            data-testid="left-element"
          />
          <input
            class="w-full bg-transparent placeholder-neutral-500 outline-0"
            data-testid="input:input"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col"
    >
      <div
        class="flex flex-1 flex-row items-center gap-x-3 rounded-xl border border-neutral-200 bg-white p-3 px-5"
        data-testid="input:container"
      >
        <div
          data-testid="left-element"
        />
        <input
          class="w-full bg-transparent placeholder-neutral-500 outline-0"
          data-testid="input:input"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
